File: src/io/innofang/proxy/example/DynamicProxy.java
package io.innofang.proxy.example;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

/**
 * Created by Inno Fang on 2017/4/8.
 *
 * Dynamic proxy
 */
public class DynamicProxy implements InvocationHandler {

    private Object object;

    public DynamicProxy(Object object) {
        this.object = object;
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        return method.invoke(object, args);
    }
}

File: src/io/innofang/proxy/example/kotlin/DynamicProxy.kt
package io.innofang.proxy.example.kotlin

import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

/**
 * Created by Inno Fang on 2017/9/3.
 */
class DynamicProxy(private val obj: Any) : InvocationHandler {

    override fun invoke(proxy: Any?, method: Method?, args: Array<out Any>?): Any? {
        return method!!.invoke(obj, *(args ?: arrayOfNulls<Any>(0)))
    }
}